README.rst

```diff
@@ -387,6 +387,55 @@ include any additional headers.
         resp = session.send(prepped)
         assert resp.text == "hello world"
 
+Response Registry
+---------------------------
+
+By default, ``responses`` module will search all registered``Response`` objects and
+return a match. If only one ``Response`` is registered, the registry is kept unchanged.
+However, if multiple matches are found for the same request, then first match is returned and
+removed from registry.
+
+Such behavior is suitable for most of use cases, but to handle special conditions, you can
+implement custom registry which must follow interface of ``registries.FirstMatchRegistry``.
+Redefining the ``find`` method will allow you to create custom search logic and return
+appropriate ``Response``
+
+Example that shows how to set custom registry
+
+.. code-block:: python
+
+    import responses
+    from responses import registries
+
+
+    class CustomRegistry(registries.FirstMatchRegistry):
+        pass
+
+
+    print("Before tests:", responses.mock._get_registry())
+
+    # using function decorator
+    @responses.activate(registry=CustomRegistry)
```

Comment:
Do you find that yo need to change the registry strategy in each test? Should we only support `responses.registry` as the way to change the registry globally and `RequestsMock.registry` for a locally scoped registry when it is required?

Reply:
@markstory 
the idea here is similar to `mock.patch` where you patch each test.

in general, we can compliment this implementation with making `get` and `set` registry public methods. Exposing `registry` directly allows to set non empty registry. The question would be if we want to allow this or do we want to be more conservative and allow only empty registry that should be filled?

Reply:
let me provide a detailed example.

one of the worst things in unit testing that can happen, is a leakage of environment from one test to another
Let's take an example, based on proposed alternative implementation where one gets `.registry` attribute:
```python
import responses
from responses.registries import StrictOrderRegistry, PopFirstKeepLastRegistry

assert type(responses.mock.registry) == PopFirstKeepLastRegistry

with responses.RequestsMock() as rsps:
    rsps.registry = StrictOrderRegistry()
    assert type(rsps.registry) == StrictOrderRegistry


@responses.activate
def my_test_func():
    responses.mock.registry = StrictOrderRegistry()
    assert type(responses.mock.registry) == StrictOrderRegistry


my_test_func()
assert type(responses.mock.registry) == PopFirstKeepLastRegistry
# >>> AssertionError, since responses.mock.registry is equal to StrictOrderRegistry
```
what we see, that if anybody in some test will assign value of the registry, then it will propagate to all other tests.
we need to support identical behavior for both Context Manager and for decorator. Of course, we can revert to default registry on `__exit__`, but this will be:
1. implicit (I prefer to be explicit whenever it is possible) 
2. will cause issues when `responses.mock.registry` set outside of the function (globally), since this will be applied only to a first function executed

on top of this we have additional minor issue, user can make a typo in attribute name and still think that registry is set, while it is not. eg `responses.mock.regstry`

this leads that we need to set registry on each test, reasons for this:
1. explicit
2. allows user to precisely control behavior of tests

drawback that we need to do extra coding is eliminated with new proposal of **backwards compatible** decorator, that can take argument of registry `@responses.activate(registry=CustomRegistry)` or `registry` argument in context manager itself. Both ways should be easy to use.

Reply:
> [...] Let's take an example, based on proposed alternative implementation [...]

This is the point where I think I need to explain the intention and difference between this PR and #440.
In #440 the implementation is not advertising the registry to be used on the module level instance `responses.mock`. Instead it is implemented similar to `responses.mock.response_callback`, where all the warnings about the leakage of environment are equally valid for the current version of this library. However, in contrast to the module level `response_callback` which is just a reference to `None`, a module level reference to `registry` could make sense, assuming that the registry of the default mock is never swapped out.
So the difference is, that this PR tries to implement the registry functionality to be also used with the default mock, while #440 intents to provide the functionality only for local instances of `RequestsMock`.
Maybe it is just me who is scared about global (module level) instances, but note how much complexity is needed to restore the initial state of the default mock instance... What about a global reset function which just recreates a new instance of default mock with a well defined state? Of course, this reset function will also need to reset the global pointers to the new instance methods and attributes.